benchmarks/microbenchmarks.py

```diff
@@ -1,37 +1,151 @@
+
+class TraitA(OpTrait):
+    """An example trait."""
+
+
+class TraitB(OpTrait):
+    """An example trait."""
+
+
+@irdl_op_definition
+class HasTraitAOp(IRDLOperation):
+    """An operation which has a trait A."""
+
+    name = "has_trait_a"
+    traits = traits_def(TraitA())
+
+
+class IRTraversal:
+    """Benchmark the time to traverse xDSL IR."""
+
+    EXAMPLE_BLOCK_NUM_OPS = 1_000
+    EXAMPLE_OPS = (EmptyOp() for _ in range(EXAMPLE_BLOCK_NUM_OPS))
+    EXAMPLE_BLOCK = Block(ops=EXAMPLE_OPS)
+
+    def time_iterate_ops(self) -> None:
+        """Time directly iterating over a python list of operations.
+
+        Comparison with `for (Operation *op : /*std::vector*/ops) {`.
+        """
+        for op in IRTraversal.EXAMPLE_OPS:
+            assert op
+
+    def time_iterate_block_ops(self) -> None:
+        """Time directly iterating over the linked list of a block's operations.
+
+        Comparison with `for (Operation &op : *block) {`.
+        """
+        for op in IRTraversal.EXAMPLE_BLOCK.ops:
+            assert op
+
+    def time_walk_block_ops(self) -> None:
+        """Time walking a block's operations.
+
+        Comparison with `block->walk([](Operation *op) {});` with no region in
+        the IR.
+        """
+        for op in IRTraversal.EXAMPLE_BLOCK.walk():
+            assert op
+
+
+class Extensibility:
+    """Benchmark the time to check interface and trait properties."""
+
+    HAS_TRAIT_A_OP = HasTraitAOp()
+
+    def time_interface_check(self) -> None:
+        """Time checking the class hierarchy of an operation.
+
+        Indirect comparison with `assert( dyn_cast<OpT>(op) )`.
+
+        This is not a direct comparison as xDSL does not use the
+        class hierarchy to express interface functionality, but is interesting
+        to compare `isinstance` with `dyn_cast` in context.
+        """
+        assert isinstance(Extensibility.HAS_TRAIT_A_OP, HasTraitAOp)
+
+    def time_trait_check(self) -> None:
+        """Time checking the trait of an operation.
+
+        Comparison with `assert( op->hasTrait<TraitT>(op) )`.
+        """
+        assert Extensibility.HAS_TRAIT_A_OP.has_trait(TraitA)
+
+    def time_trait_check_neg(self) -> None:
+        """Time checking the trait of an operation.
+
+        Comparison with `assert( ! op->hasTrait<TraitT>(op) )`.
         """
-        importlib.reload(xdsl.dialects.arith)
+        assert not Extensibility.HAS_TRAIT_A_OP.has_trait(TraitB)
 
-    def time_builtin_load(self) -> None:
-        """Time loading the `builtin` dialect."""
-        importlib.reload(xdsl.dialects.builtin)
+
+class OpCreation:
+    """Benchmark creating an operation in xDSL."""
+
+    CONSTANT_OPERATION = EmptyOp()
+
+    def time_operation_create(self) -> None:
+        """Time creating an empty operation.
+
+        Comparison with `OperationState opState(unknownLoc, "testbench.empty");
+        Operation::create(opState)`.
+        """
+        EmptyOp.create()
```

Comment:
Can we have another one where we pass a few operands / result types / regions?

Reply:
I would recommend a separate PR for this

Reply:
Fine by me